crates/driver/src/infra/blockchain/mod.rs

```diff
@@ -184,54 +185,66 @@ impl Ethereum {
-                vec![serde_json::to_value(&tx).unwrap(), "pending".into()],
-            )
-            .await?;
-        if let Some(err) = json.get("error") {
-            return Err(Error::AccessList(err.to_owned()));
-        }
-        let access_list: web3::types::AccessList =
-            serde_json::from_value(json.get("accessList").unwrap().to_owned()).unwrap();
-        Ok(access_list.into())
+        let tx = tx.into();
+
+        let gas_limit = self.inner.tx_gas_limit.try_into().map_err(|err| {
+            Error::GasPrice(anyhow!("failed to convert gas_limit to u64: {err:?}"))
+        })?;
+        let tx = tx.with_gas_limit(gas_limit);
+        let tx = match self.simulation_gas_price().await {
+            Some(gas_price) => {
+                let gas_price = gas_price.try_into().map_err(|err| {
+                    Error::GasPrice(anyhow!("failed to convert gas_limit to u128: {err:?}"))
+                })?;
+
+                tx.with_gas_price(gas_price)
+            }
+            _ => tx,
+        };
+
+        let access_list = self.web3.alloy.create_access_list(&tx).pending().await?;
+
+        Ok(access_list
+            .ensure_ok()
+            .map_err(Error::AccessList)?
+            .access_list
+            .into())
     }
 
     /// Estimate gas used by a transaction.
-    pub async fn estimate_gas(&self, tx: &eth::Tx) -> Result<eth::Gas, Error> {
-        self.web3
-            .eth()
-            .estimate_gas(
-                web3::types::CallRequest {
-                    from: Some(tx.from.into_legacy()),
-                    to: Some(tx.to.into_legacy()),
-                    value: Some(tx.value.0.into_legacy()),
-                    data: Some(tx.input.clone().into()),
-                    access_list: Some(tx.access_list.clone().into()),
-                    gas_price: self
-                        .simulation_gas_price()
-                        .await
-                        .map(IntoLegacy::into_legacy),
-                    ..Default::default()
-                },
-                Some(BlockNumber::Pending),
-            )
-            .await
-            .map(IntoAlloy::into_alloy)
-            .map(Into::into)
-            .map_err(Into::into)
+    pub async fn estimate_gas(&self, tx: eth::Tx) -> Result<eth::Gas, Error> {
+        let tx = TransactionRequest::default()
+            .from(tx.from)
+            .to(tx.to)
+            .value(tx.value.0)
+            .input(tx.input.0.into())
+            .access_list(tx.access_list.into());
+
+        let tx = match self.simulation_gas_price().await {
+            Some(gas_price) => {
+                let gas_price = gas_price.try_into().map_err(|err| {
+                    Error::GasPrice(anyhow!("failed to convert gas_limit to u128: {err:?}"))
+                })?;
+
+                tx.with_gas_price(gas_price)
+            }
+            _ => tx,
+        };
+
+        tracing::error!("estimating gas");
+
+        Ok(U256::from(
```

Comment:
The formatting looks off. Probably slightly nicer to first write the result into a variable before converting it into the final result.